### PR TITLE
Task/dss123 add channel settings

### PR DESCRIPTION
**Pull request by FChataigner**
[sc-150953]

### CODE REVIEW

File: python-clusters/create-aks-cluster/cluster.json [hunk 1]
```diff
@@ -155,6 +155,36 @@
...
+        },
+        {
+            "name": "nodeOSUpgradeChannel",
+            "label": "Node OS upgrades",
+            "description": "Determines whether and when the OS in the cluster nodes is upgraded",
+            "type": "SELECT",
+            "selectChoices": [
+                {"value":"None", "label":"No automatic upgrade"},
+                {"value":"Unmanaged", "label":"Upgrades handled by the OS"},
+                {"value":"NodeImage", "label":"Weekly upgrades of the node image"},
```
Comment (thtrunck) [inline python-clusters/create-aks-cluster/cluster.json:181]: nitpicking, is there a reason to not follow the order of the doc (SecurityPatch above NodeImage?)

File: python-clusters/create-aks-cluster/cluster.json [hunk 2]
```diff
@@ -155,6 +155,36 @@
             "mandatory": false,
             "defaultValue": "Standard"
         },
+        {
+            "name": "upgradeChannel",
+            "label": "Cluster upgrades",
+            "description": "Determines whether and when the cluster is automatically upgraded based on new K8S or AKS releases",
+            "type": "SELECT",
+            "selectChoices": [
+                {"value":"none", "label":"No automatic upgrade"},
```
Comment (thtrunck) [inline python-clusters/create-aks-cluster/cluster.json:164]: Nitpicking here too, don't we want the order from the slowest to the fastest?
none, patch, stable, rapid, node-image?

**Review by thtrunck — CHANGES_REQUESTED** [review]
There is a big question regarding support of python 27/3.6.
We will also not correctly reflect the current status of already created cluster, I don't know if that's an issue.
And I'm also wondering if we want a macro or something to just update the current cluster configuration. Asking customer to recreate the cluster just to update the policy from none to unmanaged may not be great.
Finally I have no idea what happens if you don't have the feature preview on your subscription and you try to do that? Maybe I can just unregister the feature on our account, I will try that

File: python-clusters/create-aks-cluster/cluster.json [hunk 3]
```diff
@@ -155,6 +155,36 @@
...
+        },
+        {
+            "name": "nodeOSUpgradeChannel",
+            "label": "Node OS upgrades",
+            "description": "Determines whether and when the OS in the cluster nodes is upgraded",
+            "type": "SELECT",
+            "selectChoices": [
+                {"value":"None", "label":"No automatic upgrade"},
+                {"value":"Unmanaged", "label":"Upgrades handled by the OS"},
+                {"value":"SecurityPatch", "label":"Weekly upgrades handled by AKS - security patches only"},
```
Comment (thtrunck) [inline python-clusters/create-aks-cluster/cluster.json:181]:
```suggestion
                {"value":"SecurityPatch", "label":"Weekly upgrades handled by AKS - security patches only (preview)"},
```
Maybe we could had a preview flag to let them know they need the FF activated. The error is explicit but it's unclear the ff only impact this channel.

**Review by thtrunck — CHANGES_REQUESTED** [review]
With 37+ python if I don't have the FF with Stable/securityPatch on my subscription I got
image
(it works with other node upgrade settings). And the failure is quite fast so it's not too bad

I think the Exception message aren't super actionable. With the plugin selection if they are in 27/36 it's likely they don't have python 37+ so asking them upgrade will probably block them for a while.
Can we either tell them in the error that they can upgrade the codeenv or use the default (and tell them what's the default in their case) or just have a warning and fallback to the default.
I don't know what's the best option.

File: python-clusters/create-aks-cluster/cluster.json (hunk 3 above)
Comment (FChataigner) [reply python-clusters/create-aks-cluster/cluster.json]: agree. Not much more we can do anyway, I don't want to go all the way and probe which preview features are active (you'll need elevated permissions for that anyway, I bet)

**Review by amandineslx — APPROVED** [review]
OK for me

### Tested

* deploy DSS 12.2.2 in Elastic fleet with FM in Azure
* import plugin build from branch in PR (commit `5de4675270b3fd301ba5c1287a25db8bb51c8111`)
* change the plugin code env to one using Python 3.6
* create a cluster
  * **Node pools**
    * **Machine type=Standard_B8ms;Default number of nodes=1;Min number of nodes=1;Max number of nodes=2;Availability zones=unticked**
  * **Service CIDR=10.1.0.0/16;DNS IP=10.1.0.10;Load balancer SKU=Standard;Cluster upgrades=No automatic upgrade;Node OS upgrades=Upgrades handled by the OS**
  * **SAVE**
  * **START**
    * :white_check_mark: see error
    ```
    Failed to start cluster : <class 'Exception'> : Python SDK is too old to pass node OS upgrade channel setting. Create a new plugin code env with python >=3.7
    ```
* change the plugin code env to one using Python 3.7
* stop cluster
* update cluster
  * **Cluster upgrades=Latest patch version on currently used minor;Node OS upgrades=Upgrades handled by the OS**
  * **SAVE**
* :white_check_mark: start cluster
* :white_check_mark: check cluster configuration in cluster page
```
"auto_upgrade_profile": {
  "upgrade_channel": "patch",
  "node_os_upgrade_channel": "Unmanaged"
},
```
* stop cluster
* update cluster
  * **Cluster upgrades=Latest patch release on next to latest minor;Node OS upgrades=Weekly upgrades handled by AKS**
  * **SAVE**
* :white_check_mark: start cluster
* :white_check_mark: check cluster configuration in cluster page
```
"auto_upgrade_profile": {
  "upgrade_channel": "stable",
  "node_os_upgrade_channel": "SecurityPatch"
},
```
* stop cluster
* update cluster
  * **Cluster upgrades=Latest patch release on latest minor;Node OS upgrades=Weekly upgrades by AKS**
  * **SAVE**
* :white_check_mark: start cluster
* :white_check_mark: check cluster configuration in cluster page
```
"auto_upgrade_profile": {
  "upgrade_channel": "rapid",
  "node_os_upgrade_channel": "Unmanaged"
},
```
* stop cluster
* update cluster
  * **Cluster upgrades=Latest patch release on latest minor;Node OS upgrades=Weekly upgrades of the node image**
  * **SAVE**
* :white_check_mark: start cluster
* :white_check_mark: check cluster configuration in cluster page
```
"auto_upgrade_profile": {
  "upgrade_channel": "rapid",
  "node_os_upgrade_channel": "NodeImage"
},
```
* stop cluster
* update cluster
  * **Cluster upgrades=Latest node image version**
  * **SAVE**
* :white_check_mark: start cluster
* :white_check_mark: check cluster configuration in cluster page
```
"auto_upgrade_profile": {
  "upgradeChannel": "node-image",
  "node_os_upgrade_channel": "NodeImage"
},
```

Note:
* command to check cluster status would be: : `az aks show --resource-group asouilleux-fm --name asouilleux-cluster`

**Review by thtrunck — APPROVED** [review]
LGTM.
I haven't retested everything but I did some random testing with python 2.7/3.6/3.9. The support of python 2.7/3.6 is not perfect but the errors message appears quickly enough and are actionable.
The error when the az feature isn't active on the subscription is also clear (+from the dropdown it's clear that this is a preview).

**Pull request merged**